Add Strapi authentication

dahuang37 opened the pull request:
adds a new file `strapi-api.ts` for implementing API calls to Strapi.
It encapsulate the auth implementation in its class and provide a few
example function for calling to strapi endpoint.

Bonus:
- refactored the auth to google-api, which also encapsulate the auth
functionality in the file and only expose the API functionality.
- refactored handler to use the new google-api and strapi-api.
- Fixes the issue with parsing admin reports from Google API. 

Testing:
works locally with miniflare. additional testing is required for
cloudflare worker.

Resolves #4 

Signed-off-by: Daniel Huang <danielhuang37@outlook.com>

dahuang37 committed:
Add Strapi authentication

adds a new file `strapi-api.ts` for implementing API calls to Strapi.
It encapsulate the auth implementation in its class and provide a few
example function for calling to strapi endpoint.

Bonus:
- refactored the auth to google-api, which also encapsulate the auth
functionality in the file and only expose the API functionality.
- refactored handler to use the new google-api and strapi-api.

Testing:
works locally with miniflare. additional testing is required for
cloudflare worker.

Signed-off-by: Daniel Huang <danielhuang37@outlook.com>

chungthuang commented on src/strapi-api.ts:
```diff
@@ -0,0 +1,57 @@
+            "identifier": username,
+            "password": password
+        }
+        let strapiResponse = await fetch("http://localhost:1337/auth/local", {
```
Should we change this to `STRAPIBASEURL`?

dahuang37 replied:
updated.

dahuang37 commented on src/strapi-api.ts:
```diff
@@ -0,0 +1,57 @@
+        const response = await fetch(url, {
+            method: 'GET',
+            headers: {
+            'cookie': this.token
```
indent

dahuang37 committed:
fixup! Add Strapi authentication

Signed-off-by: Daniel Huang <danielhuang37@outlook.com>

chungthuang commented on src/strapi-api.ts:
```diff
@@ -0,0 +1,58 @@
+
+        var username: string = await cron_worker.get("strapi_username") as string;
+        var password: string = await cron_worker.get("strapi_password") as string;
+        if (!username || !password) {
```
Can we comment that this is for dev environment?

dahuang37 replied:
added comment

dahuang37 committed:
Updates comments for strapi-api

dahuang37 committed:
Add fix for fetching google reports

Previously fetching google reports were failing because it was
missing subject emaill as part of the parameter.
This commit adds the fix and example for fetching it.

chungthuang commented on src/google-api.ts:
```diff
@@ -0,0 +1,103 @@
+
+export class GoogleAPI {
+    authenticator: GoogleJwtAuthenticater;
+    adminScopes = "https://www.googleapis.com/auth/calendar https://www.googleapis.com/auth/admin.directory.group https://www.googleapis.com/auth/admin.directory.group.member https://www.googleapis.com/auth/admin.reports.audit.readonly https://www.googleapis.com/auth/admin.directory.user.security" ;
```
Can this be an array? It's easier to read that way.

dahuang37 replied:
the scope has to be a single string and not an array. Added comments to explain the issue.

dahuang37 committed:
add comments about google api and avoid null iteration on reports
